src/Amalgam/entity/Entity.cpp

```diff
@@ -896,74 +889,91 @@ void Entity::SetRoot(EvaluableNode *_code, bool allocated_with_entity_enm, Evalu
 	SetRoot(new_code, true, metadata_modifier, write_listeners);
 }
 
-void Entity::AccumRoot(EvaluableNodeReference accum_code, bool allocated_with_entity_enm, EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier, std::vector<EntityWriteListener *> *write_listeners)
+void Entity::AccumRoot(EvaluableNodeReference accum_code, bool allocated_with_entity_enm,
+	EvaluableNodeManager::EvaluableNodeMetadataModifier metadata_modifier,
+	std::vector<EntityWriteListener *> *write_listeners)
 {
-	if( !(allocated_with_entity_enm && metadata_modifier == EvaluableNodeManager::ENMM_NO_CHANGE))
+	if(!(allocated_with_entity_enm && metadata_modifier == EvaluableNodeManager::ENMM_NO_CHANGE))
 		accum_code = evaluableNodeManager.DeepAllocCopy(accum_code, metadata_modifier);
 
-	bool accum_has_labels = EvaluableNodeTreeManipulation::DoesTreeContainLabels(accum_code);
+	auto [new_labels, no_label_collisions] = EvaluableNodeTreeManipulation::RetrieveLabelIndexesFromTree(accum_code);
 
 	EvaluableNode *previous_root = evaluableNodeManager.GetRootNode();
-	EvaluableNodeReference new_root = AccumulateEvaluableNodeIntoEvaluableNode(EvaluableNodeReference(previous_root, true), accum_code, &evaluableNodeManager);
 
-	//need to check if still cycle free as it may no longer be
-	EvaluableNodeManager::UpdateFlagsForNodeTree(new_root);
+	//before accumulating, check to see if flags will need to be updated
+	bool node_flags_need_update = false;
+	if(previous_root == nullptr)
+	{
+		node_flags_need_update = true;
+	}
+	else
+	{
+		//need to update node flags if new_root is cycle free, but accum_node isn't
+		if(previous_root->GetNeedCycleCheck() && (accum_code != nullptr && !accum_code->GetNeedCycleCheck()))
+			node_flags_need_update = true;
+
+		//need to update node flags if new_root is idempotent and accum_node isn't
+		if(previous_root->GetIsIdempotent() && (accum_code != nullptr && accum_code->GetIsIdempotent()))
+			node_flags_need_update = true;
+	}
+
+	//accum, but can't treat as unique in case any other thread is accessing the data
+	EvaluableNodeReference new_root = AccumulateEvaluableNodeIntoEvaluableNode(
+		EvaluableNodeReference(previous_root, false), accum_code, &evaluableNodeManager);
 
 	if(new_root != previous_root)
 	{
-		//keep reference for current root (mainly in case a new node was added if the entity were previously empty)
+		//keep reference for current root before setting and freeing
 		evaluableNodeManager.KeepNodeReference(new_root);
-
 		evaluableNodeManager.SetRootNode(new_root);
-
-		//free current root reference
 		evaluableNodeManager.FreeNodeReference(previous_root);
 	}
 
-	size_t num_root_labels_to_update = 0;
-	if(new_root != nullptr)
-		num_root_labels_to_update = new_root->GetNumLabels();
+	//optimistically create references for the new labels, delete them if find collisions
+	string_intern_pool.CreateStringReferences(new_labels, [](auto l) { return l.first; });
```

Comment:
Why do you create these refs up front and then del when it collides? Why not just try to inset and if it does then create the ref? Overhead of continually calling ref creation all at once vs in the loop?